protocol/lavasession/provider_session_manager_test.go

```diff
@@ -118,6 +164,115 @@ func TestHappyFlowPSM(t *testing.T) {
 	require.Equal(t, sps.PairingEpoch, epoch1)
 }
 
+func TestHappyFlowBadgePSM(t *testing.T) {
+	badgeSession := &BadgeSession{}
+	// init test
+	psm, sps := prepareBadgeSession(t, context.Background(), badgeSession, 0)
+
+	// on session done successfully
+	err := psm.OnSessionDone(sps, relayNumber)
+
+	// validate session done data
+	require.Nil(t, err)
+	require.Equal(t, sps.LatestRelayCu, uint64(0))
+	require.Equal(t, sps.CuSum, relayCu)
+	require.Equal(t, sps.BadgeUserData.UsedComputeUnits, relayCu)
+	require.Equal(t, sps.SessionID, sessionId)
+	require.Equal(t, sps.RelayNum, relayNumber)
+	require.Equal(t, sps.PairingEpoch, epoch1)
+}
+
+func TestHappyFlowBadgePSMMultipleRoutines(t *testing.T) {
+	// Num of goroutines to run
+	numRoutines := 1000
+
+	// A channel to track goroutine completion
+	done := make(chan struct{})
+
+	var totalRelayCU uint64 // total served CU for onSessionDone
+	badgeSessionsUsedCU := make([]*ProviderSessionsEpochData, numRoutines)
+
+	// Testing in multiple goroutines
+	for i := 0; i < numRoutines; i++ {
+		index := i // Create a new variable and assign loop variable's value to it
+		go func() {
+			// Decrement the wait group counter when the goroutine finishes
+			defer func() {
+				done <- struct{}{}
+			}()
+
+			// Create a unique BadgeSession for each goroutine
+			badgeSession := &BadgeSession{}
+
+			// Initialize the test
+			psm, sps := prepareBadgeSession(t, context.Background(), badgeSession, index)
+			// Put session into global array
+			badgeSessionsUsedCU[index] = sps.BadgeUserData
+
+			// Randomly determine test success or failure with equal probability
+			rand.Seed(time.Now().UnixNano())
+			isSuccess := rand.Intn(2) == 0
+
```

Comment:
missing some random sleep to make the cuncurrency really kick in otherwise it can finish too fast, please add random sleep of 0, 1ms 2ms

Reply:
Added here: 0afedb82b69ec0a1a7fde38bb9b0b2d9df187e9a